pkg/storage/tsdb/bucketindex/index.go

```diff
@@ -0,0 +1,195 @@
+package bucketindex
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	IndexFilename = "bucket-index.json"
+	IndexVersion1 = 1
+
+	SegmentsFormatUnknown       = ""
+	SegmentsFormat1Based6Digits = "1b6d"
+)
+
+// Index contains all known blocks and markers of a tenant.
+type Index struct {
+	// Version of the index format.
+	Version int `json:"version"`
+
+	// List of complete blocks (partial blocks are excluded from the index).
+	Blocks []*Block `json:"blocks"`
+
+	// List of block deletion marks.
+	BlockDeletionMarks []*BlockDeletionMark `json:"block_deletion_marks"`
+
+	// UpdatedAt is a unix timestamp (seconds precision) of when the index has been updated
+	// (written in the storage) the last time.
+	UpdatedAt int64 `json:"updated_at"`
+}
+
+// Block holds the information about a block in the index.
+type Block struct {
+	// Block ID.
+	ID ulid.ULID `json:"block_id"`
+
+	// MinTime and MaxTime specify the time range all samples in the block are in (millis precision).
+	MinTime int64 `json:"min_time"`
+	MaxTime int64 `json:"max_time"`
+
+	// SegmentsFormat and SegmentsNum stores the format and number of chunks segments
+	// in the block, if they match a known pattern. We don't store the full segments
+	// files list in order to keep the index small. SegmentsFormat is empty if segments
+	// are unknown or don't match a known format.
+	SegmentsFormat string `json:"segments_format"`
```

Comment:
I'd suggest to omit empty segments_format and segments_num fields.

Reply:
Good point!